### ISSUE TITLE
Defragmentation for partitioned InnoDB table

### ISSUE THREAD

**Issue by SunguckLee** [facebook/mysql-5.6#8]
Hi Facebook.
Really thanks for your awesome features.

I really interested in defragmentation of InnoDB table.
But for the huge table, we can't split defragmentation job.
Yes for the index level But if PRIMARY KEY itself is huge, we have to run defragmentation whole day.

So, I think defragmentation job can be splitted by partition level if the table is partitioned.
But current facebook version of mysql doesn't support for the partitioned table.

I think it can be implement easily (Just think) and I want to try.
Before implementation, You can also implement it easily (you are expert than me at least ^^), but you didn't.
I thought there's enough reason you did't implement it. Please share the reason to me ?

Really Thanks.

**Comment by rongrong** [comment]
Hi Sunguck,

Thanks for your feedback and we are glad that you are interested in the work.

There are a few reasons:
1.  The feature is developed targeting our largest tier, which does not have many partitioned table, so partitioned table is not a problem for us, as a result we never thought about that. (OK, this is the real reason.)
2.  The speed of defragmentation itself is not as important as its performance impact on production work load. Instead of optimizing on how to finish up defragmenting everything fast, we tried to make defragmentation not affecting normal work load by rate limiting it’s speed, and control how many index are being defragmented in parallel. It might take a even longer time to defragment, but if nobody will notice anything while it’s running it doesn’t matter that much.

How big is your table (or primary index)?

Thanks!
Rongrong

**Comment by SunguckLee** [comment]
Hi rongrong.
Really thanks for your detailed explanation.

The biggest table which I am targeting is about 400GB, and this table is partitioned (hash partition).
But now we can't convert it to non-partitioned table for defragmentation. ^^
So I thought that defragmentation features which can be applied to partitioned-table.

Anyway, I will try it.
I will ask you again if there's problem.

Have a nice day.
Matt.

**Comment by steaphan-fb-com** [comment]
Cool.  Let us know how it goes with another comment here.

Or, if you have any more problems, please open a new issue with the details.

**Comment by SunguckLee** [comment]
Hi Rongrong and Speaphan.

I added some code for supporting defragmentation of partitioned innodb table.
I want to the code change will be merged to facebook version of mysql.
We also want to use facebook version mysql on our service if requirement is meet ^^.
(There's some ALTER TABLE syntax change, but if you mind that you can ignore it)

Regards,
Matt.
